src/ripple/app/tx/impl/InvariantCheck.cpp

```diff
@@ -358,6 +358,101 @@ AccountRootsNotDeleted::finalize(
 
 //------------------------------------------------------------------------------
 
+void
+AccountRootsDeletedClean::visitEntry(
+    bool isDelete,
+    std::shared_ptr<SLE const> const& before,
+    std::shared_ptr<SLE const> const&)
+{
+    if (isDelete && before && before->getType() == ltACCOUNT_ROOT)
+        accountsDeleted_.emplace_back(before);
+}
+
+bool
+AccountRootsDeletedClean::finalize(
+    STTx const& tx,
+    TER const result,
+    XRPAmount const,
+    ReadView const& view,
+    beast::Journal const& j)
+{
+    // This list should include all of the keylet functions that take a single
+    // AccountID parameter, except nftpage_max, because it is exercised below
+    // using the NFT page lookup logic. (nftpage_min is included because it
+    // should _never_ be present.)
+    using fType = std::function<Keylet(AccountID const& id)>;
+    static std::array<fType, 4> const keyletfuncs{
+        &keylet::account,
+        &keylet::ownerDir,
+        &keylet::signers,
+        &keylet::nftpage_min,
+        //&keylet::nftpage_max
+    };
+
+    // Always check for objects in the ledger, but to prevent differing
+    // transaction processing results, however unlikely, only fail if the
+    // feature is enabled. Enabled, or not, though, a fatal-level message will
+    // be logged
+    bool const enforce = view.rules().enabled(featureInvariantsV1_1);
+
+    auto const objectExists = [&view, enforce, &j](auto const& keylet) {
+        if (auto const sle = view.read(keylet))
+        {
+            // Finding the object is bad
+            auto const typeName = [&sle]() {
+                auto item =
+                    LedgerFormats::getInstance().findByType(sle->getType());
+
+                if (item != nullptr)
+                    return item->getName();
+                return std::to_string(sle->getType());
+            }();
+
+            JLOG(j.fatal())
+                << "Invariant failed: account deletion left behind a "
+                << typeName << " object";
+            assert(enforce);
```

Comment:
Won't this assert mean that it doesn't log if the amendment is enabled?

Reply:
It should log - the logging is single-threaded last I looked, so it'll finish before the assert fires.

Reply:
I was mostly confused by the `assert` statement before the `return` statement.

Reply:
Does this invariant check not crash if the assert fails/if the amendment isn't enabled? Or is that what you meant by the logging happens before the assert fires?

Reply:
> Does this invariant check not crash if the assert fails/if the amendment isn't enabled? Or is that what you meant by the logging happens before the assert fires?

If the amendment isn't enabled, and a problem is detected
1. The message will be written to the log.
2. In a debug build (or a release build with asserts enabled), the assert will fire, which will crash rippled.
3. In a release build:
    1. The `objectExists` lambda will return `true`, indicating the object exists.
    2. The callers all have `&& enabled` after the call to `objectExists`, so they will not return a failure to the transaction engine. It'll look like the invariant was fine.

The important thing here is that production servers typically run release builds. Anybody running a debug build will typically know what they're doing, and want to hit `assert`s if something goes wrong.

I wanted to implement it this way so that developers (who are more likely to run debug builds) will run into that `assert` if they make a change that breaks the invariant (hopefully while running unit tests!), regardless of whether the amendment has been enabled on the ledger.

Reply:
Ah, the part I was missing was that `assert`s only run on a debug build. Thanks!